pyrobosim/pyrobosim/navigation/astar_grid.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python3
+import math
+from pyrobosim.utils.pose import Pose
+from pyrobosim.utils.motion import Path
+from pyrobosim.navigation.occupancy_grid import occupancy_grid_from_world
+
+
+class Node:
+    def __init__(self, x, y, g, h, parent=None) -> None:
+        self.x = x
+        self.y = y
+        self.g = g
+        self.h = h
+        self.parent = parent
+
+    @property
+    def f(self):
+        return self.g + self.h
+
+    def __eq__(self, __o):
+        """Equality definition for a Node"""
+        return self.x == __o.x and self.y == __o.y
+
+    def __repr__(self) -> str:
+        return f"({self.x}, {self.y}) : {self.f}\n"
+
+
+class AstarGrid:
+    """Occupancy grid based A-star planner"""
+
+    def __init__(self, w, resolution=0.05, inflation_radius=0.0) -> None:
+        self.grid = occupancy_grid_from_world(
+            w, resolution=resolution, inflation_radius=inflation_radius
+        )
+        self.world = w
+        self.resolution = resolution
+        self.inflation_radius = inflation_radius
+        self.candidates = []  # Node
+        self.visited = []  # (x, y)
+        self.start = None
+        self.goal = None
+
+    def _reset(self):
+        """Resets the state of data used by the algorithm"""
+        self.grid = occupancy_grid_from_world(
+            self.world,
+            resolution=self.resolution,
+            inflation_radius=self.inflation_radius,
+        )
+        self.candidates.clear()
+        self.visited.clear()
+        self.goal = None
+        self.start = None
+        self.latest_path = []
+
+    def _heuristic(self, p1, p2):
+        """
+        Computes a heuristc between 2 points
+        :param p1: Source point
+        :type p1: Tuple (x, y)
+        :param p2: destination point
+        :type p2: Tuple (x, y)
+        """
+        return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
+
+    def _expand(self, node):
+        """Generates and adds free neighbours to exploration candidates"""
+        motions = [(-1, 0), (1, 0), (0, 1), (0, -1), (-1, 1), (-1, -1), (1, 1), (1, -1)]
+        motion_costs = [1, 1, 1, 1, 1.414, 1.414, 1.414, 1.414]
+        for i in range(len(motions)):
+            _x = node.x + motions[i][0]
+            _y = node.y + motions[i][1]
+            if (_x, _y) not in self.visited and not self.grid.is_occupied(_x, _y):
```

Comment:
Because this is a list, it's expensive to check this "visited" condition since it's linear search. Easiest fix is to use a set instead, which makes this check constant-time with hashing. Try it out and compare timings!

Reply:
Set :  0.0064640045166015625 seconds
List :  0.007136344909667969 seconds

There is a slight speedup, it might be more significant for larger grid sizes or longer  and complex paths.

Reply:
Interesting. Try make the resolution super small and see if the difference is better?

Reply:
Switch to the PriorityQueue , the time was cut down to half ,  0.003673553466796875 seconds